### PR TITLE
feat: reduce allocs in query sharder

### PR DESCRIPTION
**Pull request by javiermolinar**
**What this PR does**:
Reduce memory allocations when sharding the query. Similar perf improvement as #3901  

**Checklist**
- [X] Tests updated
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by joe-elliott — DISMISSED** [review]
Looks like there's some minor conflicts to resolve. Once that's done we'll merge.

File: modules/frontend/traceid_sharder.go [hunk 1]
```diff
@@ -75,17 +76,18 @@ func (s *asyncTraceSharder) buildShardedRequests(ctx context.Context, parent *ht
 	for i := 0; i < len(s.blockBoundaries); i++ {
 		reqs[i] = parent.Clone(ctx)
 
-		q := reqs[i].URL.Query()
+		qb := api.NewQueryBuilder(reqs[i].URL.RawQuery)
```
[the remark this answers is not in the document]
Comment (javiermolinar) [reply modules/frontend/traceid_sharder.go:79]: Something like this?

// It returns an URL with query params
func BuildQuery(url string, args ...string) (string, error) {
}

[the remark this answers is not in the document]
Comment (javiermolinar) [reply modules/frontend/traceid_sharder.go]: I think it's better to pass a map. The map keys can be reused to reduce allocations.

[the remark this answers is not in the document]
Comment (javiermolinar) [reply modules/frontend/traceid_sharder.go]: Gotcha, I've done the same but with the map and as a generic helper. I think it's nice to have it. I've moved the method to the commonplace and used the same structure

**Review by joe-elliott — APPROVED** [review]

**Pull request merged**